File: tier2_intermediate/stack/problems/p155_min_stack.c
/*
 * P155: Min Stack (Medium)
 * https://leetcode.com/problems/min-stack/
 * Topics: Stack, Design
 *
 * Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.
 * Implement the MinStack class:
 * You must implement a solution with O(1) time complexity for each function.
 *
 * Example 1:
 *     Input
 *     ["MinStack","push","push","push","getMin","pop","top","getMin"]
 *     [[],[-2],[0],[-3],[],[],[],[]]
 *
 *     Output
 *     [null,null,null,null,-3,null,0,-2]
 *
 * Hint: stdin: n, then n blocks of (op_name, value_count, values...)
 *       stdout: one line per result, "null" or integer
 */

#include "io.h"
#include <stdlib.h>
#include <string.h>

static int st_data[30000], mn_data[30000];
static int st_sz, mn_sz;

static void ms_init(void) {
    abort();
}

static void ms_push(int val) {
    abort();
}

static void ms_pop(void) {
    abort();
}

static int ms_top(void) {
    abort();
}

static int ms_get_min(void) {
    abort();
}

int main(void)
{
    int cn; int *ca = read_ints(&cn); int n = ca[0]; free(ca);
    ms_init();
    for (int i = 0; i < n; i++) {
        char *op = read_line();
        int cc; int *ca = read_ints(&cc); int count = ca[0]; free(ca);
        int val = 0;
        if (count > 0) {
            int dummy;
            int *vals = read_ints(&dummy);
            val = vals[0];
            free(vals);
        }

        if (strcmp(op, "push") == 0) {
            ms_push(val);
            write_string("null");
        } else if (strcmp(op, "pop") == 0) {
            ms_pop();
            write_string("null");
        } else if (strcmp(op, "top") == 0) {
            write_int(ms_top());
        } else if (strcmp(op, "getMin") == 0) {
            write_int(ms_get_min());
        } else {
            write_string("null");
        }
        free(op);
    }
    return 0;
}

File: tier2_intermediate/stack/problems/p155_min_stack.rs
/*
 * P155: Min Stack (Medium)
 * https://leetcode.com/problems/min-stack/
 * Topics: Stack, Design
 *
 * Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.
 * You must implement a solution with O(1) time complexity for each function.
 *
 * Hint: stdin: n, then n blocks of (op_name, value_count, values...)
 *       stdout: one line per result, "null" or integer
 */

use wasm_libs::*;

struct MinStack {
    st: Vec<i32>,
    mn: Vec<i32>,
}

impl MinStack {
    fn new() -> Self {
    todo!();
}
    fn push(&mut self, val: i32) {
    todo!();
}
    fn pop(&mut self) {
    todo!();
}
    fn top(&self) -> i32 {
    todo!();
}
    fn get_min(&self) -> i32 {
    todo!();
}
}

fn main() {
    let n = read_int();
    let mut ms = MinStack::new();
    for _ in 0..n {
        let op = read_line();
        let count = read_int();
        let val = if count > 0 { read_ints()[0] } else { 0 };

        if op == "push" {
            ms.push(val);
            write_string("null");
        } else if op == "pop" {
            ms.pop();
            write_string("null");
        } else if op == "top" {
            write_int(ms.top());
        } else if op == "getMin" {
            write_int(ms.get_min());
        } else {
            write_string("null");
        }
    }
    std::process::exit(0);
}
